testing_barrier.jl

```diff
@@ -0,0 +1,141 @@
+################################################
+# sIPOPT test
+################################################
+
+using JuMP
+using SparseArrays
+using LinearAlgebra
+using Ipopt
+# using MadNLP
+# using KNITRO
+
+include("nlp_utilities.jl")
+include("nlp_utilities_test.jl")
+
+
+test_compute_optimal_hess_jacobian()
+
+test_compute_derivatives()
+
+test_compute_derivatives_Finite_Diff()
+
+
+#######################
+# Test Bilevel (WIP)
+#######################
+
+using BilevelJuMP, Ipopt
+
+function constraint_upper!(model, x, y)
+    @constraints(model, begin
+        x <= 5
+        y <= 8
+        y >= 0
+    end)
+end
+
+function obj_upper!(model, x, y)
+    @objective(model, Min, 3x + y)
+end
+
+function build_upper!(model, x, y)
+    obj_upper!(model, x, y)
+    constraint_upper!(model, x, y)
+end
+
+function constraint_lower!(model, x, y; slack = zeros(4))
+    @constraints(model, begin
+        x +  y + slack[1] <= 8
+        4x +  y + slack[2] >= 8
+        2x +  y + slack[3] <= 13
+        2x - 7y + slack[4] <= 0
+    end)
+end
+
+function obj_lower!(model, x, y, slack)
+    @objective(model, Min, -x + 10 * sum(s^2 for s in slack))
+end
+
+function build_lower!(model, x, y, slack)
+    obj_lower!(model, x, y, slack)
+    constraint_lower!(model, x, y, slack = slack)
+end
+
+### Test BilevelJuMP ###
+using BilevelJuMP
+
+model = BilevelModel(Ipopt.Optimizer, mode = BilevelJuMP.ProductMode(1e-5))
+
+@variable(Lower(model), x_b)
+@variable(Upper(model), y_b)
+
+build_upper!(Upper(model), x_b, y_b)
+build_lower!(Lower(model), x_b, y_b, zeros(4))
+optimize!(model)
+
+objective_value(model) # = 3 * (3.5 * 8/15) + 8/15 # = 6.13...
+value(x_b) # = 3.5 * 8/15 # = 1.86...
+value(y_b) # = 8/15 # = 0.53...
+
+### Test DiffOpt Non-Convex ###
+
+model_upper = Model(Ipopt.Optimizer)
+model_lower = Model(Ipopt.Optimizer)
+
+@variable(model_upper, y)
+@variable(model_upper, x_star)
+@variable(model_upper, x_aux)
+@variable(model_lower, x)
+@variable(model_lower, y_p ∈ MOI.Parameter(1.0))
+@variable(model_lower, slack[1:4])
+
+build_upper!(model_upper, x_star, y)
+constraint_lower!(model_upper, x_aux, y)
+build_lower!(model_lower, x, y_p, slack)
+primal_vars = [x; slack]
+params = [y_p]
+evaluator, cons = create_evaluator(model_lower; x=[primal_vars; params])
+
+# Define `f(y) = x_star` & `f'(y)`
+function f(y_val)
+    set_parameter_value(y_p, y_val)
+    optimize!(model_lower)
+    @assert is_solved_and_feasible(model_lower)
+    return value(x)
+end
+
+function ∇f(y_val)
+    @assert value(y_p) == y_val
+    Δs, sp = compute_sensitivity(evaluator, cons, [0.001]; primal_vars=primal_vars, params=params)
```

Comment:
> I assume the hard-coded [0.001] is because we need to assume some step length in order to decide if we need to use fix-and-relax? 

Yes, which the zero Δp makes me wonder if there is a problem with the chosen tolerance. I will investigate now.

> Is there a way to do this without a hard-coded step size? 

Not when considering the possibility of fix and relax. What we can do is to give the option for the user to just ignore the fix and relax.

Reply:
The tolerance was too tight and I was checking one of the sensitivities on the wrong side.

Reply:
now should be fixed but I will add more individual tests for each sensitivity.